### PR TITLE
Make federate sharding possible in metrics-collector by using workers

### PR DESCRIPTION
**Pull request by saswatamcode**
Attempting to implement federate sharding by sharding the match rules (both passed as file and flag) between a number of workers. 

If the code looks messy, that is because it is, there is a lot of unused and scattered logic throughout metrics-collector. I have tried to refactor a lot of this

### CODE REVIEW

File: operators/multiclusterobservability/api/shared/multiclusterobservability_shared.go
Comment (philipgough) [inline operators/multiclusterobservability/api/shared/multiclusterobservability_shared.go:47]: nit: Workers is the number of workers that work in parallel ...   sounds more natural than the way this is written

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (thibaultmg) [inline collectors/metrics/cmd/metrics-collector/main.go:540]: I agree, most of what is done in the metrics collector is already setup in the endpoint-operator when generating the ServiceMonitors. The only missing labels are `managementcluster` and `managementclusterID`. They can be easily added.

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (thibaultmg) [inline collectors/metrics/cmd/metrics-collector/main.go:595]: I'm not sure there is much to win here, given that it will be replaced by MCOA. In my view, it's not worth the effort. It's good enough.

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:540]: Yup, we can think of a way to add it as a relabel config on Prom, using MCO. But probably in later PRs.

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:595]: Hmm, MCOA would still use metrics-collector, right? Or are you saying we won't be reporting status from here in MCOA anymore?

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (thibaultmg) [inline collectors/metrics/cmd/metrics-collector/main.go:199]: What is the purpose for exposing these flags?

Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:199]: e2e tests in the future. I already have one like so, but to add it here, would need some go.mod changes which we can tackle in later PRs.

```go
package e2e_test

import (
	"fmt"
	"os"
	"strconv"
	"testing"
	"time"

	"github.com/efficientgo/e2e"
	e2edb "github.com/efficientgo/e2e/db"
	e2einteractive "github.com/efficientgo/e2e/interactive"
	e2emon "github.com/efficientgo/e2e/monitoring"
	e2eobs "github.com/efficientgo/e2e/observable"
	"github.com/thanos-io/thanos/test/e2e/e2ethanos"

	"github.com/efficientgo/core/backoff"

	"github.com/efficientgo/core/testutil"
)

func TestMetricsCollectorE2E(t *testing.T) {
	e, err := e2e.NewDockerEnvironment("metrics-col")
	testutil.Ok(t, err)
	t.Cleanup(e.Close)

	m, err := e2emon.Start(e)
	testutil.Ok(t, err)

	prom := e2edb.NewPrometheus(e, "prom2")
	receive1 := e2ethanos.NewReceiveBuilder(e, "receiver-1").WithImage("quay.io/thanos/thanos:v0.35.1").WithIngestionEnabled().Init()

	testutil.Ok(t, e2e.StartAndWaitReady(prom, receive1))

	collector := NewMetricsCollector(e, "collector-1", MetricsCollectorOptions{
		From:     prom.InternalEndpoint("http"),
		To:       receive1.InternalEndpoint("remote-write"),
		Matchers: []string{`{__name__="up"}`, `{__name__="go_gc_duration_seconds"}`, `{__name__="go_goroutines"}`, `{__name__="go_memstats_alloc_bytes"}`, `{__name__="go_memstats_heap_alloc_bytes"}`, `{__name__="go_memstats_heap_idle_bytes"}`, `{__name__="go_memstats_heap_inuse_bytes"}`, `{__name__="go_memstats_heap_objects"}`, `{__name__="go_memstats_heap_released_bytes"}`, `{__name__="go_memstats_heap_sys_bytes"}`, `{__name__="go_memstats_last_gc_time_seconds"}`, `{__name__="go_memstats_lookups_total"}`, `{__name__="go_memstats_mallocs_total"}`, `{__name__="go_memstats_sys_bytes"}`, `{__name__="go_threads"}`, `{__name__="process_cpu_seconds_total"}`, `{__name__="process_open_fds"}`, `{__name__="process_resident_memory_bytes"}`, `{__name__="process_start_time_seconds"}`, `{__name__="process_virtual_memory_bytes"}`},
		// Matchers: []string{`{__name__="up"}`},
		LogLevel: "info",
		Interval: "2s",
		Workers:  4,
	})

	testutil.Ok(t, e2e.StartAndWaitReady(collector))

	time.Sleep(10 * time.Second)

	testutil.Ok(t, receive1.WaitSumMetricsWithOptions(e2emon.GreaterOrEqual(10), []string{"prometheus_tsdb_head_series"}, e2emon.WaitMissingMetrics()))

	query1 := e2edb.NewThanosQuerier(
		e,
		"query1",
		[]string{
			receive1.InternalEndpoint("grpc"),
		},
		e2edb.WithImage("quay.io/thanos/thanos:v0.35.1"),
	)
	testutil.Ok(t, e2e.StartAndWaitReady(query1))
	testutil.Ok(t, query1.WaitSumMetricsWithOptions(e2emon.Equals(1), []string{"thanos_store_nodes_grpc_connections"}, e2emon.WaitMissingMetrics()))
	testutil.Ok(t, e2einteractive.OpenInBrowser(fmt.Sprintf("http://%s", prom.Endpoint("http"))))
	testutil.Ok(t, e2einteractive.OpenInBrowser(fmt.Sprintf("http://%s", query1.Endpoint("http"))))
	testutil.Ok(t, m.OpenUserInterfaceInBrowser())
	testutil.Ok(t, e2einteractive.RunUntilEndpointHit())

}

type MetricsCollectorOptions struct {
	From     string
	To       string
	Matchers []string
	LogLevel string
	Interval string
	Workers  int
}

func NewMetricsCollector(e e2e.Environment, name string, o MetricsCollectorOptions) *e2eobs.Observable {
	f := e.Runnable(name).WithPorts(map[string]int{"metrics": 9002}).Future()

	args := e2e.BuildArgs(map[string]string{
		"--from":                     "http://" + o.From,
		"--to-upload":                "http://" + o.To + "/api/v1/receive",
		"--log-level":                o.LogLevel,
		"--interval":                 o.Interval,
		"--disable-status-reporting": "true",
		"--disable-hypershift":       "true",
		"--worker-number":            strconv.Itoa(o.Workers),
	})

	for _, matcher := range o.Matchers {
		args = append(args, "--match="+matcher)
	}

	return e2eobs.AsObservable(f.Init(wrapWithDefaults(e2e.StartOptions{
		Image:   "example.com/metrics-collector:v0.0.1",
		Command: e2e.NewCommandWithoutEntrypoint("metrics-collector", args...),
	})), "metrics")
}

var defaultBackoffConfig = backoff.Config{
	Min:        300 * time.Millisecond,
	Max:        600 * time.Millisecond,
	MaxRetries: 50,
}

func wrapWithDefaults(opt e2e.StartOptions) e2e.StartOptions {
	if opt.User == "" {
		opt.User = strconv.Itoa(os.Getuid())
	}
	if opt.WaitReadyBackoff == nil {
		opt.WaitReadyBackoff = &defaultBackoffConfig
	}
	return opt
}

```

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (thibaultmg) [inline collectors/metrics/cmd/metrics-collector/main.go:302]: There is probably some code we can extract in a specific function from `initShardedConfigs`. It mutates the options objects, doing the same processing 3 times.

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (thibaultmg) [inline collectors/metrics/cmd/metrics-collector/main.go:755]: `opt.Labels` already exists at this stage no? It's been generated in `initShardedConfigs`. We could iterate directly on it. And why are we setting the `cluster` and `clusterID` at this stage and not in the other function?
I might be lost

Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:755]: So this runMultiWorkers function isn't really used anywhere in main flow, this is part of the simulation testing flow. I just moved the function, so it diffs this way, but I haven't actually modified anything in this function,

https://github.com/stolostron/multicluster-observability-operator/blob/main/collectors/metrics/cmd/metrics-collector/main.go#L442

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:302]: Yeah you're right, just didn't want to change even more things on sensitive code path

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:755]: As for why it is the way it is, I don't really know, but I wanted to remove this simulation testing, it is mixed in with business logic and is not really a great simulation of anything. Can do in later PRs

**Review by thibaultmg — APPROVED** [review]
Awesome! Made some comments but no blocker.

/lgtm

**Comment by saswatamcode** [comment]
/hold

Holding for a bit so others can take a look

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (thibaultmg) [reply collectors/metrics/cmd/metrics-collector/main.go:595]: The metrics collector should be replaced by the prometheusAgent. Nothing is set in stone though.

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (thibaultmg) [reply collectors/metrics/cmd/metrics-collector/main.go:755]: Ack no worries, I did not pay too much attention between what was already there, and what you added. We can leave it as is.

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:595]: > The metrics collector should be replaced by the prometheusAgent. Nothing is set in stone though.

Yup me and @moadz discussed this, but I don't think it would be feasible yet, as you can't downsample, so would end up sending all raw 30s scrape interval data

File: collectors/metrics/pkg/metricsclient/metricsclient_test.go
Comment (moadz) [inline collectors/metrics/pkg/metricsclient/metricsclient_test.go:26]: `http.Dial` is deprecated, `http.DialContext` instead

File: collectors/metrics/pkg/status/status.go
Comment (moadz) [inline collectors/metrics/pkg/status/status.go:73]: `forwarder.go` we use noop reporter by default - why the need for a status reporter at all there?

File: operators/multiclusterobservability/api/shared/multiclusterobservability_shared.go
Comment (moadz) [inline operators/multiclusterobservability/api/shared/multiclusterobservability_shared.go:58]: nit: Is it clear on the addon spec that workers corresponds to prometheus federate workers and not something else?

**Review by moadz — APPROVED** [review]
LGTM - awesome work Saswata! 

I don't, however, think it's appropriate to merge this day before code-freeze. Feels like there is substantial testing risk associated with this change and I wouldn't want to compromise the GA date. We can discuss this further in sync/offline.

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:358]: Ah yes missed that one

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:399]: This is perfscale simulations I believe, nothing I added, it existed originally. 
https://github.com/stolostron/multicluster-observability-operator/tree/main/tools/simulator/metrics-collector

We really should drop this

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:444]: Not exactly afaik.
Collectrules are rules that when evaluated and exceed a particular threshold, we just add a matcher to collect new metric
Recording Rules are actual recording rules that are evaluated in metrics-collector and sent as new series.

File: collectors/metrics/pkg/forwarder/forwarder.go
Comment (saswatamcode) [reply collectors/metrics/pkg/forwarder/forwarder.go:418]: I think this should be exponential backoff retry not quite what this is.
Right now, if a request fails, regardless of interval you set, it will just wait a full minute and try again.

File: collectors/metrics/pkg/forwarder/forwarder_test.go
Comment (saswatamcode) [reply collectors/metrics/pkg/forwarder/forwarder_test.go:169]: This was set in kind of wrong way tbh, but got handled using a else branch that checks for the UNIT_TEST var directly in business logic, which is even more fishy. I don't think we are passing it as upload CA exactly.

I can instead just edit test struct instead

File: collectors/metrics/pkg/metricsclient/metricsclient_test.go
Comment (saswatamcode) [reply collectors/metrics/pkg/metricsclient/metricsclient_test.go:26]: Ack, will update, this was old code

File: collectors/metrics/pkg/status/status.go
Comment (saswatamcode) [reply collectors/metrics/pkg/status/status.go:73]: Forwarder is in charge of federate and remote writing that federate. So that is where we report status.

Earlier it always reported status using kubeAPI, but for local testing might not need that. So I made the status reporting kind of optional.

File: operators/multiclusterobservability/api/shared/multiclusterobservability_shared.go
Comment (saswatamcode) [reply operators/multiclusterobservability/api/shared/multiclusterobservability_shared.go:58]: I'll add a better description

**Review by moadz — APPROVED** [review]
LGTM pending commnets and the hold

File: collectors/metrics/cmd/metrics-collector/main.go
Comment (saswatamcode) [reply collectors/metrics/cmd/metrics-collector/main.go:595]: Ah yeah in that case it is, but still promagent would end up doing single large /federate call

File: collectors/metrics/pkg/forwarder/forwarder_test.go
Comment (saswatamcode) [reply collectors/metrics/pkg/forwarder/forwarder_test.go:217]: This is the feature we dropped basically which is worker reconfigure. This test was testing exactly that

**Review by philipgough — APPROVED** [review]
/lgtm

**Pull request merged**
